### ISSUE TITLE
RuntimeError: Unable to find a valid cuDNN algorithm to run convolution

### ISSUE THREAD

**Issue by Michael-Li-Bryant** [gml16/rl-medical#13]
Hi @gml16 , I'm trying to learn sth from this project.
I just cloned it to my PC and used the instruction of Evaluating with 8 C-MARL agents.
However, I met an error saying that'RuntimeError: Unable to find a valid cuDNN algorithm to run convolution'.
I'm sure that I'm using the same environment as the file 'environment.yml' offered.
Btw, I reinstalled cuDNN 7.6.5 but it didn't work.
Also, I've looked for several solutions saying that decreasing the batch size could be helpful.
I changed the batch size from 64 to 32 but the same problem occurred.
So where could the problem be at?
Maybe it's a very basic problem since this is my first time to apply a project on my own.I'm really desperate for the training.
Looking forward to your reply.

**Comment by gml16** [comment]
Hi Michael, thanks for the question. Assuming you do have a GPU and Cuda is properly installed, the main factor for VRAM growth comes from the buffer size (the buffer is stored on the GPU). I’d recommend trying ti set `--init_memory_size` and `--memory_size` to a very small number such as 10. Feel free to also try to set the batch size to 1. If that doesn’t work could you also send the output of running `nvidia-smi` on your system please?

**Comment by Michael-Li-Bryant** [comment]
Thanks for your apply. I'm actually using an RTX 3090 so I don't think there could be sth wrong with the buffer size. As the program running, the buffer size will come to a maximum of about 800MiB then that error would appear. I will just follow your advice to set --init_memory_size and --memory_size to a smaller number and I'll try my best. Thanks to your contribution. Also further questions could appear since I'm not confident enough to solve the problem completely. So maybe I'll ask you more about this project. Thx a lot.
Best wishes.

**Comment by Michael-Li-Bryant** [comment]
Sorry to bother so many times. Actually I'm wondering sth more.
I just want to know about your GPU devices and CUDA version while running this project.

**Comment by gml16** [comment]
No worries at all, there is more information about the hardware in my report in [section 5.2](https://gml16.github.io/projects/mastersthesis.pdf#page36). 

“We used CUDA version 10.0.130, Torch v1.4, GPU cards had 12GB RAM and were either Nvidia Tesla or Nvidia GeForce GTX Titan Xp, 24-core/48 thread Intel Xeon CPUs with 256GB RAM.”

RTX 3090 should be more than enough. Maybe newer Cuda versions are not compatible. 

Let me know if the error persists.

**Comment by Michael-Li-Bryant** [comment]
Actually I'm following your suggestions to restart my training.
However, the same error occured once more.
Here's the Traceback:

Traceback (most recent call last):
  File "DQN.py", line 234, in <module>
    evaluator.play_n_episodes(fixed_spawn=args.fixed_spawn)
  File "/home/Michael_Bryant/rl-medical/src/evaluator.py", line 45, in play_n_episodes
    score, start_dists, q_values, info = self.play_one_episode(render, fixed_spawn=fixed_spawn[j])
  File "/home/Michael_Bryant/rl-medical/src/evaluator.py", line 86, in play_one_episode
    acts, q_values = predict(obs_stack)
  File "/home/Michael_Bryant/rl-medical/src/evaluator.py", line 74, in predict
    q_vals = self.model.forward(inputs).detach().squeeze(0)
  File "/home/Michael_Bryant/rl-medical/src/DQNModel.py", line 189, in forward
    x = self.conv0(x)
  File "/home/Michael_Bryant/.conda/envs/rl-medical/lib/python3.8/site-packages/torch/nn/modules/module.py", line 722, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/Michael_Bryant/.conda/envs/rl-medical/lib/python3.8/site-packages/torch/nn/modules/conv.py", line 566, in forward
    return F.conv3d(input, self.weight, self.bias, self.stride,
RuntimeError: Unable to find a valid cuDNN algorithm to run convolution

And I'm using the command like this:
python DQN.py --task eval --load 'data/models/BrainMRI/CommNet8agents.pt' --files 'data/filenames/image_files.txt' 'data/filenames/landmark_files.txt' --batch_size 1 --memory_size 10 --init_memory_size 10 --file_type brain --landmarks 13 14 0 1 2 3 4 5 --model_name "CommNet"

Here's my output of the command nvidia-smi.I'm using the server of our school so maybe the output is quite complicated.

+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.60.11    Driver Version: 525.60.11    CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  Off  | 00000000:1A:00.0 Off |                  N/A |
| 30%   38C    P8    29W / 350W |  11732MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA GeForce ...  Off  | 00000000:1B:00.0 Off |                  N/A |
| 52%   55C    P2   121W / 350W |    258MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  NVIDIA GeForce ...  Off  | 00000000:3D:00.0 Off |                  N/A |
| 43%   49C    P2   113W / 350W |   4860MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  NVIDIA GeForce ...  Off  | 00000000:3E:00.0 Off |                  N/A |
| 30%   38C    P8    25W / 350W |   2110MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   4  NVIDIA GeForce ...  Off  | 00000000:88:00.0 Off |                  N/A |
| 64%   60C    P2   144W / 350W |  14570MiB / 24576MiB |     28%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   5  NVIDIA GeForce ...  Off  | 00000000:89:00.0 Off |                  N/A |
| 52%   55C    P2   138W / 350W |   3468MiB / 24576MiB |     17%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   6  NVIDIA GeForce ...  Off  | 00000000:B1:00.0 Off |                  N/A |
| 51%   55C    P2   113W / 350W |   7512MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   7  NVIDIA GeForce ...  Off  | 00000000:B2:00.0 Off |                  N/A |
| 54%   57C    P2   126W / 350W |   1508MiB / 24576MiB |     11%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A     10475      C   python                            254MiB |
|    0   N/A  N/A     19216      C   python                            254MiB |
|    0   N/A  N/A     21833      C   ...onda/envs/py37/bin/python    10940MiB |
|    0   N/A  N/A     30303      C   python                            254MiB |
|    1   N/A  N/A     27293      C   python                            256MiB |
|    2   N/A  N/A     29717      C   .../voxelmorph/bin/python3.8     4856MiB |
|    4   N/A  N/A     25491      C   .../voxelmorph/bin/python3.8     4856MiB |
|    4   N/A  N/A     26514      C   .../voxelmorph/bin/python3.8     4856MiB |
|    4   N/A  N/A     29153      C   .../voxelmorph/bin/python3.8     4856MiB |
|    5   N/A  N/A     30303      C   python                           3462MiB |
|    6   N/A  N/A     10475      C   python                           7502MiB |
|    7   N/A  N/A     19216      C   python                           1500MiB |
+-----------------------------------------------------------------------------+

I've also checked the command conda list and I've selected the following information:
pytorch                   1.6.0           py3.8_cuda10.2.89_cudnn7.6.5_0    pytorch
cudatoolkit               10.2.89              hfd86e86_1
cudnn                     7.6.5                cuda10.2_0
python                    3.8.5                h7579374_1

What should I do next to avoid this error?
Looking forward to your reply.

**Comment by Vathsan** [comment]
Hi Michael, this error can occur for multiple reasons. Since you get the error even for running the evaluation, I do not think reducing the batch size would help. There may be some issues with the library versions. You could try a couple of things and see if that works for you.
1. Updating the [nvidia driver](https://huggingface.co/CompVis/stable-diffusion-v1-4/discussions/37)
2. Updating the cuda and cudnn version. I was able to run the same command with the following library versions on RTX 3050 TI and Tesla P100
```
nvidia-cublas-cu11==11.10.3.66
nvidia-cuda-cupti-cu11==11.7.101
nvidia-cuda-nvrtc-cu11==11.7.99
nvidia-cuda-runtime-cu11==11.7.99
nvidia-cudnn-cu11==8.5.0.96
nvidia-cufft-cu11==10.9.0.58
nvidia-curand-cu11==10.2.10.91
nvidia-cusolver-cu11==11.4.0.1
nvidia-cusparse-cu11==11.7.4.91
nvidia-nccl-cu11==2.14.3
nvidia-nvtx-cu11==11.7.91
```
I have attached the [requirements.txt](https://github.com/gml16/rl-medical/files/11254244/requirements.txt) file, if you wanna take a look.

**Comment by Michael-Li-Bryant** [comment]
Hi Vathsan, thanks to your help, I finally got able to start my training.
Since I am a junior in applying a project, I want to consult some advice when building my own environment.
I used just the exact "environment.yml" the project offered, so how could the problem occur?
Also, if sth wrong really happens, how could I just build a proper environment for the project just as you did?
Thanks a lot. Looking forward to your guidance.
